### PR TITLE
refactor: squad posting to increment squad total posts

### PR DESCRIPTION
**Pull request by sshanzel**
As the RFC has mentioned, we should update the mutation for creating new posts related to Squad.

There was no mention of deleting posts, but I assumed we should reduce the count by 1. Feel free to mention if we should not make any changes to it.

Notable changes here:
- Created a generic function for saving changes to posts, this is to allow us to streamline the updating of flags (maybe we could have done the increment on the new post worker, but I might have missed the discussion around it).
- Created a generic function for deleting a post, to do the same above.

MI-322 #done

### CODE REVIEW

File: src/entity/posts/utils.ts
Comment (sshanzel) [inline src/entity/posts/utils.ts:379]: Nothing was changed with the object we passed as `post`, just indentation due to being inside an object.

File: src/common/post.ts
Comment (sshanzel) [inline src/common/post.ts:135]: Nothing was changed with the object we passed as `post`, just indentation due to being inside an object.

File: src/entity/posts/utils.ts
Comment (capJavert) [inline src/entity/posts/utils.ts:65]: Might be out of loop but did we not say we are going to do this with triggers? This way counts will be updated no matter from where the post is deleted (API, script, psql etc.) we do similar for `upvotes` and `downvotes` for posts.

FYI @DragosIorgulescu

Comment (sshanzel) [reply src/entity/posts/utils.ts:65]: Yeah, DB triggers are a good option as well. Looking forward to the reply as the ticket mentioned the mutation 👀

Comment (sshanzel) [reply src/entity/posts/utils.ts:65]: I just saw the note about DB triggers being an option. I can maybe try and see how we can incorporate it. However, it only mentions the `totalPosts` to be part of it and the others would be within a cron job.

Comment (sshanzel) [reply src/entity/posts/utils.ts:65]: The RFC highlights updating the mutations, we can maybe reopen a ticket if we ever decide to finalize using DB triggers. @capJavert feel free to make the final call, I am cool either way and willing to make the changes if necessary.

**Comment by sshanzel** [comment]
@capJavert I pushed some changes towards using triggers. Would love to hear your initial thoughts. This is only for total posts count. I will be adding the triggers for the other 2 properties. Do you maybe want them in a separate migration file or I can keep them all in the same one?

File: src/migration/1714633817175-SquadTotalPostsTrigger.ts
Comment (capJavert) [inline src/migration/1714633817175-SquadTotalPostsTrigger.ts:16]: I think better to just do `AND s.type = 'squad'` in UPDATE statement, that will avoid select query in each trigger execution if I see correctly?

File: src/migration/1714633817175-SquadTotalPostsTrigger.ts
Comment (capJavert) [inline src/migration/1714633817175-SquadTotalPostsTrigger.ts:19]: Never seen this `array['totalPosts']` syntax. It actually returns `{totalPosts}` which is what `jsonb_set` requires. I would prefer to just use string, seems more simple if there is not anything else I am missing?

File: src/migration/1714633817175-SquadTotalPostsTrigger.ts
Comment (capJavert) [reply src/migration/1714633817175-SquadTotalPostsTrigger.ts:16]: UPDATE that does not modify any rows should be faster then SELECT + OPTIONALLY UPDATE.

Comment (capJavert) [reply src/migration/1714633817175-SquadTotalPostsTrigger.ts:16]: Actually we can also use `WHEN (NEW.type = 'squad'` in trigger definition, that should be the best.

**Comment by capJavert** [comment]
@sshanzel 

>  I will be adding the triggers for the other 2 properties. Do you maybe want them in a separate migration file or I can keep them all in the same one?

Whatever you prefer. 👍

File: src/migration/1714633817175-SquadTotalPostsTrigger.ts
Comment (sshanzel) [reply src/migration/1714633817175-SquadTotalPostsTrigger.ts:16]: But the type here is the post, so it will never be a `squad`.

File: src/migration/1714633817175-SquadTotalPostsTrigger.ts
Comment (sshanzel) [reply src/migration/1714633817175-SquadTotalPostsTrigger.ts:19]: The `array['totalPosts']` just looks more object-oriented and modern, I have no hard preference, so either way works for me.

File: src/migration/1714633817175-SquadTotalPostsTrigger.ts
Comment (capJavert) [reply src/migration/1714633817175-SquadTotalPostsTrigger.ts:16]: Yeah but you are doing UPDATE query on `source` where you can do a `WHERE` filter on its type.

Comment (sshanzel) [reply src/migration/1714633817175-SquadTotalPostsTrigger.ts:16]: Oh, right. Forgot about that. Yes, we can do that.

File: src/migration/1714633817175-SquadTotalPostsTrigger.ts
Comment (capJavert) [inline src/migration/1714633817175-SquadTotalPostsTrigger.ts:35]:
```suggestion
            SET     flags = jsonb_set(flags, '{totalPosts}', to_jsonb(GREATEST(0, COALESCE(CAST(flags->>'totalPosts' AS INTEGER), 0) - 1)))
```

Just to make sure it handles no value and not go into negative with `GREATEST`.

**Review by capJavert — APPROVED** [review]
One edge case, but good to go after it, nice work!

**Review by kopancek — APPROVED** [review]
I like the trigger implementation, agree with Ante on safeguarding the delete though.

File: src/migration/1714633817175-SquadTotalPostsTrigger.ts
Comment (sshanzel) [reply src/migration/1714633817175-SquadTotalPostsTrigger.ts]: Yep. Makes sense to catch the edge-case.

**Comment by sshanzel** [comment]
@capJavert @kopancek @nensidosari all triggers to update the stats (totalViews, totalPosts, totalUpvotes) should now be available.

File: src/migration/1714633817175-SquadTotalPostsTrigger.ts
Comment (rebelchris) [inline src/migration/1714633817175-SquadTotalPostsTrigger.ts:17]: We decided on this just for performance? (Just meant maybe not a bad thing to have for all sources, but also not needed now)

File: src/migration/1714662136871-SquadTotalUpvotesTrigger.ts
Comment (rebelchris) [inline src/migration/1714662136871-SquadTotalUpvotesTrigger.ts:37]: More of a question @capJavert is there no way we can only attach trigger to set column?

Comment (rebelchris) [reply src/migration/1714662136871-SquadTotalUpvotesTrigger.ts:37]: Alternative, we listen to the same thing here, wondering if it's "cheaper" to simply just do all with 1 trigger?

**Review by rebelchris — APPROVED** [review]
Looks pretty good, just some questions none blocking.

File: src/migration/1714633817175-SquadTotalPostsTrigger.ts
Comment (capJavert) [inline src/migration/1714633817175-SquadTotalPostsTrigger.ts:36]: Why two `jsonb_set`? Should it not the second one be separate trigger? That is how the rest were done.

File: src/migration/1714633817175-SquadTotalPostsTrigger.ts
Comment (capJavert) [inline src/migration/1714633817175-SquadTotalPostsTrigger.ts:42]: Also not sure if we need to decrement views when post is deleted, views are still there and happened, no matter if post is deleted.

File: src/migration/1714662136871-SquadTotalUpvotesTrigger.ts
Comment (capJavert) [reply src/migration/1714662136871-SquadTotalUpvotesTrigger.ts:37]: Hmm I think each set of triggers should listen on the source table:
- totalUpvotes triggers to `user_post` table
- totalPosts triggers `post` table (as it is now)
- totalView triggers on `view` table

File: src/migration/1714662136871-SquadTotalUpvotesTrigger.ts
Comment (capJavert) [inline src/migration/1714662136871-SquadTotalUpvotesTrigger.ts:25]: But you don't know if upvotes for how much the upvotes changed. For example upvotes could change from 8 to 10. You will increment by 1, when in reality it should add two upvotes.

Comment (capJavert) [reply src/migration/1714662136871-SquadTotalUpvotesTrigger.ts:25]: See my comment https://github.com/dailydotdev/daily-api/pull/1866#discussion_r1588928823

File: src/migration/1714633817175-SquadTotalPostsTrigger.ts
Comment (capJavert) [reply src/migration/1714633817175-SquadTotalPostsTrigger.ts:42]: Sorry, asked in discussion, did not see it before https://dailydotdev.slack.com/archives/C069XNYMDGV/p1714669514654279

File: src/migration/1714662136871-SquadTotalUpvotesTrigger.ts
Comment (capJavert) [reply src/migration/1714662136871-SquadTotalUpvotesTrigger.ts:37]: As to listening to specific table that might not be possible due to https://dailydotdev.slack.com/archives/C069XNYMDGV/p1714669514654279

Comment (rebelchris) [reply src/migration/1714662136871-SquadTotalUpvotesTrigger.ts:37]: Disregard my comment on combining and column listening :)

Comment (sshanzel) [reply src/migration/1714662136871-SquadTotalUpvotesTrigger.ts:37]: Yeah, it would have been ideal to listen to each those tables. The only downside is that we have to fetch the source id from the parent entity which is the post (for view and user_post).

File: src/migration/1714633817175-SquadTotalPostsTrigger.ts
Comment (sshanzel) [reply src/migration/1714633817175-SquadTotalPostsTrigger.ts:36]: This is to set two properties of the same flags property.

File: src/migration/1714662136871-SquadTotalUpvotesTrigger.ts
Comment (sshanzel) [reply src/migration/1714662136871-SquadTotalUpvotesTrigger.ts:25]: Would it? Do we not send an upvote one at a time per user and they will run on their own even if two users hit the upvote at the same time?

File: src/migration/1714662136871-SquadTotalUpvotesTrigger.ts
Comment (sshanzel) [reply src/migration/1714662136871-SquadTotalUpvotesTrigger.ts:37]: If we are cool doing for that, I'd be glad to attach the triggers closest to where the action is being recorded.

File: src/migration/1714633817175-SquadTotalPostsTrigger.ts
Comment (capJavert) [reply src/migration/1714633817175-SquadTotalPostsTrigger.ts:36]: Yeah sure, I just think move it to separate trigger I think so only field is updated in each trigger, as you did for others.

File: src/migration/1714662136871-SquadTotalUpvotesTrigger.ts
Comment (capJavert) [reply src/migration/1714662136871-SquadTotalUpvotesTrigger.ts:37]: Yeah, lets close this discussion and do it on post as you did, and in regards to slack thread.

File: src/migration/1714633817175-SquadTotalPostsTrigger.ts
Comment (capJavert) [reply src/migration/1714633817175-SquadTotalPostsTrigger.ts:42]: Resolved in https://github.com/dailydotdev/daily-api/pull/1866#discussion_r1589028400

File: src/migration/1714662136871-SquadTotalUpvotesTrigger.ts
Comment (capJavert) [reply src/migration/1714662136871-SquadTotalUpvotesTrigger.ts:25]: Most of the time yes, and if you listened to `user_post` that would be the case since each new row means upvote count changed by at most 1 (+ or -). But since you are looking at accumulator on post you have to take it into account.

File: src/migration/1714662136871-SquadTotalUpvotesTrigger.ts
Comment (capJavert) [inline src/migration/1714662136871-SquadTotalUpvotesTrigger.ts:37]: I am missing the trigger that on post soft deletion you decrement upvotes from squad, right?

File: src/migration/1714662136871-SquadTotalUpvotesTrigger.ts
Comment (capJavert) [reply src/migration/1714662136871-SquadTotalUpvotesTrigger.ts:25]: Between I think in this trigger you can actually use `WHEN NEW.upvotes <> OLD.upvotes` to trigger only when upvotes count change.

File: src/migration/1714633817175-SquadTotalPostsTrigger.ts
Comment (capJavert) [reply src/migration/1714633817175-SquadTotalPostsTrigger.ts:36]: Also @sshanzel what happens if post gets un-deleted. I am fine in not covering that case because it was not mentioned in RFC as well but we have to make a note of it somewhere.

File: src/migration/1714662136871-SquadTotalUpvotesTrigger.ts
Comment (sshanzel) [reply src/migration/1714662136871-SquadTotalUpvotesTrigger.ts:37]: Yep. Since we plan on separating the triggers for each properties, I will also add a separate one for this.

File: src/migration/1714662136871-SquadTotalUpvotesTrigger.ts
Comment (sshanzel) [reply src/migration/1714662136871-SquadTotalUpvotesTrigger.ts:25]: Oh, nice. Didn't know you can actually specify that.

File: src/migration/1714633817175-SquadTotalPostsTrigger.ts
Comment (sshanzel) [reply src/migration/1714633817175-SquadTotalPostsTrigger.ts:17]: Yeah, we can probably do that. That would just mean we will recalculate the stats for all sources on launch to have its merits.

Comment (capJavert) [reply src/migration/1714633817175-SquadTotalPostsTrigger.ts:17]: Does RFC say anything about this? If not let's keep it as is and we can see next week when Drgaos is back where do we want to go. I don't have strong opinions if we want to have it on all sources. I would expect this kind of aggregations could be done by clickhouse in the future so that is the reason against calculating now for every source.

Comment (rebelchris) [reply src/migration/1714633817175-SquadTotalPostsTrigger.ts:17]: Yeah would indeed roll out like this for now.

File: src/migration/1714633817175-SquadTotalPostsTrigger.ts
Comment (sshanzel) [reply src/migration/1714633817175-SquadTotalPostsTrigger.ts:42]: Yeah, sounds good to me. That was also the reason I asked it, I wanted the deduction to not happen. Let's do it.

File: src/migration/1714633817175-SquadTotalPostsTrigger.ts
Comment (sshanzel) [reply src/migration/1714633817175-SquadTotalPostsTrigger.ts]: We are not doing any deduction on the stats when a post gets deleted. Should make our lives easier on that edge case.

File: src/migration/1714633817175-SquadTotalPostsTrigger.ts
Comment (sshanzel) [reply src/migration/1714633817175-SquadTotalPostsTrigger.ts:17]: Awesome. Thank you both.

File: __tests__/posts.ts
Comment (capJavert) [inline __tests__/posts.ts:2041]: We have triggers in separate test file `__tests__/triggers/post.ts`, sorry for missing this in earlier review.

File: __tests__/posts.ts
Comment (capJavert) [inline __tests__/posts.ts:3559]: Can we keep this test as is and then just add new tests after. Old tests should pass.

**Review by capJavert — COMMENTED** [review]
Lee thanks for work on this 💪 , just a minor comments on tests, other then that looks ok to me.

File: __tests__/posts.ts
Comment (capJavert) [inline __tests__/posts.ts:3511]: Same as https://github.com/dailydotdev/daily-api/pull/1866#discussion_r1589596497

File: __tests__/posts.ts
Comment (sshanzel) [reply __tests__/posts.ts:2041]: Ah, sure, we can move it.

File: __tests__/posts.ts
Comment (sshanzel) [reply __tests__/posts.ts:3559]: Yeah, it is probably better to keep old ones untouched.

**Review by capJavert — APPROVED** [review]

**Pull request merged**
